document decision process (#4528)
ElektraInitiative/libelektra: doc/decisions/decision_process.md

# Decision Process

## Problem

Simply discussing in an issue and then doing an implementation is okay for non-substantial changes.
But substantial decisions must be made transparently and in a participative way.

## Constraints

- All relevant information about decisions must be within Elektra's repository.
- All decisions must go through at least two review rounds, with a merge in between.
- At least two people need to be in favor of the decision in both rounds.
- [Documentation guidelines](/doc/contrib/documentation.md) apply.

## Assumptions

- People want to be informed about or even participate in what Elektra looks like in the future.

## Considered Alternatives

> - Issues like https://issues.libelektra.org/4521

Reviewer: While I agree that issue #4521 isn't the best for readability, that is purely because it is actually multiple issues combined into one. If it were split up into seperate issues it wouldn't be a problem at all IMHO.

Reviewer 2: The reason I made a single issue, was 1) because it was easier to do 2) it's meant as an overview of what's left for new-backend. We can of course split it into multiple issues, if there is a need for discussions. Some of the question might not need much discussion (e.g. `elektraModulesInit` error needed basically no discussion)

Author: 321a85749

- [''using patterns to capture architectural decisions''](https://dl.acm.org/doi/10.1109/MS.2007.124)
- [arc42 decisions](http://docs.arc42.org/section-9/), and
- [RFCs in rust-lang](https://github.com/rust-lang/rfcs).

Following subsections describe all steps a decision might run through.
Only two of them are mandatory.

### Drafts

The first step is to create a PR with:

- **one** decision copied from [template.md](template.md), where at least the "Problem" and "Considered Alternatives" are filled out.
- a link from [README.md](README.md) from the "Drafts" section to this decision.

### In Discussion

This step is mandatory.

Here you must ensure:

- consistency with other decisions
- links from/to related decisions are created
- problem, constraint and assumptions are fully described and sound
- there are several considered alternatives, each with rationale and implication
- decision, rationale and implications is **not** yet filled out if there are people arguing for different options (to keep the discussion unbiased)

### In Progress

- You must include all further alternative proposals made in the "Considered Alternatives" section.

> - Now it is allowed to have the ??? from the previous round's preferred decision in the decision section.

Reviewer 2: Adding the "???" into the file doesn't make the sentence better ;)

Author: 2eca969c2

But substantial decisions must be made transparently and in a participative way.

## Constraints

- All relevant information about decisions must be within Elektra's repository.
- All decisions must go through at least two review rounds, with a merge in between.
- At least two people need to be in favor of the decision in both rounds.
- [Documentation guidelines](/doc/contrib/documentation.md) apply.
- During the decision process, the PRs constantly get updated:
  - Make changes as new commits to the pull request.
  - Questions in the PRs are answered by updating the PR and answering with the commit id.
  - As generally recommended in Elektra, do not squash after they are visible on the pull request.
  - Rebase only if the decision was already accepted and has a merge conflict.
- For reviewers:
  - Prefer to directly give suggestions how to change sentences.
  - General points should be asked in the root of "Conversation" and not at random sentences in the review.

## Assumptions

- People want to be informed about or even participate in what Elektra looks like in the future.
- We will always be able to reach an consensus.

>   We don't need a vote or a benevolent dictatorship.

Reviewer: How do we not need a vote, when the entire process is based on at least two people being in favor of the proposal in two seperate rounds? Isn't that already a vote?

- Unlike the Rust Decision process, decisions in Elektra do not have a disadvantage if they were flawed in early stages.
  Only the end results counts.

## Considered Alternatives

- Issues like https://issues.libelektra.org/4521
- PEPs: https://peps.python.org
- RFCs: https://www.ietf.org/standards/rfcs/
- Change requests: https://en.wikipedia.org/wiki/Change_request

## Decision

Decisions need to:

- be implementable within the next major release
- be according to [Elektra's goals](/doc/GOALS.md)
- first being decided upon using the decision process described here

We base our decision process and template on:

> - [''using patterns to capture architectural decisions''](https://dl.acm.org/doi/10.1109/MS.2007.124),

Reviewer: I'm not sure if basing the process onto a paper behind a paywall is such a good idead for an open-source project. Is there any way contributors outside of an academic network can access this?

Reviewer: Seems like Vienna University hosts it directly, not sure about the legality though: http://eprints.cs.univie.ac.at/2345/1/02_Using_Patterns_to_Capture.pdf

- [arc42 decisions](http://docs.arc42.org/section-9/),
- [ADR](https://adr.github.io/), and
- [RFCs in rust-lang](https://github.com/rust-lang/rfcs).

Following subsections describe all steps a decision might run through.
Only two of them are mandatory.

Explanations of the template are in [explanations.md](explanations.md).

### Drafts

The first step is to create a PR with:

- **one** decision copied from [template.md](template.md), where at least the "Problem" and "Considered Alternatives" are filled out.
- a link from [README.md](README.md) from the "Drafts" section to this decision.

### In Discussion

This step is mandatory.

Here you must ensure:

- consistency with other decisions
- links from/to related decisions are created
- problem, constraint and assumptions are fully described and sound
- there are several considered alternatives, each with rationale and implication
- decision, rationale and implications is **not** yet filled out if there are people arguing for different options (to keep the discussion unbiased)

> Here "the decision" should not have one decision but several well-described solutions.

Reviewer 2: > several well-described solutions

Where do these come from? Sounds like author has to propose multiple solution. Which would be a waste of time, if everybody already agrees on one solution.

Reviewer 2: Not quite what I meant. The issue is mostly the "well-described" part. It sounds like I have to completely think through at least two solutions, even if it is very obvious that one choice is better than the other.

Maybe something like

```suggestion
Here "the decision" should not have one decision but should describe multiple possible solutions.
The possible solutions have to be described well enough to show why the decision was made.
```

This still means we need multiple solutions, but if there is one very obvious best solution the other don't need to be described in detail.

# Decision Process

## Problem

Simply discussing in an issue and then doing an implementation is okay for non-substantial changes.
But substantial decisions must be made transparent and participative way.

## Constraints

- All relevant information about decisions must be within Elektra's repository.

> - All decisions must go through at least two review rounds, with a merge in between.

Reviewer: I guess the next point answers your question: Thwo rounds - one review after the other, by two different people.

> - All decisions must go through at least two review rounds, with a merge in between.

Reviewer 2: Oh yeah I missed the "in both rounds" part

- At least two people need to be in favor of the decision in both rounds.

## Assumptions

- People want to be informed about or even participate in what Elektra looks like in the future.

## Considered Alternatives

- Issues like https://issues.libelektra.org/4521
- PEPs: https://peps.python.org
- RFCs: https://www.ietf.org/standards/rfcs/
- Change requests: https://en.wikipedia.org/wiki/Change_request

## Decision

Decisions need to:

- be implementable within the next major release
- be according to [Elektra's goals](/doc/GOALS.md)

> - first being decided upon using the decision process described here

Reviewer: ```suggestion
- first be decided upon using the decision process described here
```

Only two of them are mandatory.

### Drafts

The first step is to create a PR with:

- **one** decision copied from [template.md](template.md), where at least the "Problem" and "Considered Alternatives" are filled out.
- a link from [README.md](README.md) from the "Drafts" section to this decision.

### In Discussion

This step is mandatory.

Here you must ensure:

- consistency with other decisions
- links from/to related decisions are created
- problem, constraint and assumptions are fully described and sound
- there are several considered alternatives, each with rationale and implication
- decision, rationale and implications is **not** yet filled out if there are people arguing for different options (to keep the discussion unbiased)

### In Progress

- You must include all further alternative proposals made in the "Considered Alternatives" section.
- Now it is allowed to have the from the previous round preferred decision in the decision section.

### Decided

- decision, rationale and implication are now filled out and fixed according the reviews
- decisions of this status usually already have an implementation PR

### Mostly Implemented

> This can be useful for decisions that need to be done for every module like plugin or library.

Author: a2427bea6

We base our decision process and template on:

- [''using patterns to capture architectural decisions''](https://dl.acm.org/doi/10.1109/MS.2007.124),
- [arc42 decisions](http://docs.arc42.org/section-9/),
- [ADR](https://adr.github.io/), and
- [RFCs in rust-lang](https://github.com/rust-lang/rfcs).

Following subsections describe all steps a decision might run through.
Only two of them are mandatory.

> Explanations of the template are in [explanations.md](explanations.md).

Reviewer 2: ```suggestion
Explanations of the template are in [template_explanations.md](template_explanations.md).
```

Maybe we even call them `TEMPLATE.md` and `EXPLANATIONS.md` to show that they are not actual decisions. Also makes the files easier to spot in a list when you're looking for them.

- The documentation links to the decision.
- The decision links to the new documentation.

> This step is finished when every reviewer approves.

Reviewer 2: See above

## Notes

> - Early discussions in issues or in discussions is not prohibited.

Reviewer 2: If that is the case, such issues should not be closed _arbitrarily_ by someone other than the person who opened the issue. Of course if it is clear to most or all participants that the problem is well defined and a decision PR is needed or that the proposal can not be implemented right now, then the issue can be closed.
